src/compiler/visitorPublic.ts

```diff
@@ -397,18 +397,18 @@ namespace ts {
             case SyntaxKind.Identifier:
                 Debug.type<Identifier>(node);
                 return factory.updateIdentifier(node,
-                    nodesVisitor(node.typeArguments, visitor, isTypeNodeOrTypeParameterDeclaration));
+                    nodesVisitor(node.typeArguments, visitor, isTypeNodeOrTypeParameterDeclaration)) as Node as T;

```

Comment:
I always feel suspicious when I see `as Node as T`. It's kind of hard during code review to tell, but what is the returned type of `nodesVisitor`?

Reply:
It's not the `nodesVisitor` that's the problem, it's `updateIdentifier`, which returns `Identifier`. But `Identifier` is not assignable to `T`, in the same way that this code errors:

```tsx
function foo<T>(x: T): T {
    if (typeof x === "number") {
        return 1234;
    }
    return x;
}
```

[Playground Link](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcA8AVAfACgB4BciGAlERogN4BQidiMwiOUAngA4CmcTeiAvIMQAiMCAC2AI04AnYSSq16ymZyggZSAIwAmAMwAWANxK6AX1OJV6zYjwmzQA)

At this point, there's nothing preventing me from doing the wrong thing here, but I think it'd be much safer to write `return factory.update...() as typeof node` rather than my silly first attempt of `as Node as T`. Then at least you get an easy-to-type way of ensuring that the two types are related.